pennylane/transforms/decompositions/two_qubit_unitary.py

```diff
@@ -0,0 +1,422 @@
+    https://link.springer.com/article/10.1007/s11128-009-0156-3
+    """
+
+    # First, write A = [[a1, a2], [-a2*, a1*]], which we can do for any SU(2) element.
+    # Then, A \otimes B = [[a1 B, a2 B], [-a2*B, a1*B]] = [[C1, C2], [C3, C4]]
+    # where the Ci are 2x2 matrices.
+    C1 = U[0:2, 0:2]
+    C2 = U[0:2, 2:4]
+    C3 = U[2:4, 0:2]
+    C4 = U[2:4, 2:4]
+
+    # From the definition of A \otimes B, C1 C4^\dag = a1^2 I, so we can extract a1
+    C14 = math.dot(C1, math.conj(math.T(C4)))
+    a1 = math.sqrt(C14[0, 0])
+
+    # Similarly, -C2 C3^\dag = a2^2 I, so we can extract a2
+    C23 = math.dot(C2, math.conj(math.T(C3)))
+    a2 = math.sqrt(-C23[0, 0])
+
+    # This gets us a1, a2 up to a sign. To resolve the sign, ensure that
+    # C1 C2^dag = a1 a2* I
+    C12 = math.dot(C1, math.conj(math.T(C2)))
+
+    if not math.allclose(a1 * np.conj(a2), C12[0, 0]):
+        a2 *= -1
+
+    # Construct A
+    A = math.stack([[a1, a2], [-math.conj(a2), math.conj(a1)]])
+
+    # Next, extract B. Can do from any of the C, just need to be careful in
+    # case one of the elements of A is 0.
+    if not math.allclose(A[0, 0], 0.0, atol=1e-6):
+        B = C1 / math.cast_like(A[0, 0], 1j)
+    else:
+        B = C2 / math.cast_like(A[0, 1], 1j)
+
+    return math.convert_like(A, U), math.convert_like(B, U)
+
+
+def _extract_su2su2_prefactors(U, V):
+    r"""U, V are SU(4) matrices for which there exists A, B, C, D such that
+    (A \otimes B) V (C \otimes D) = U. The problem is to find A, B, C, D in SU(2)
+    in an analytic and fully differentiable manner.
+
+    This decomposition is possible when U and V are in the same double coset of
+    SU(4), meaning there exists G, H in SO(4) s.t. G (Edag V E) H = (Edag U
+    E). This is guaranteed here by how V was constructed using the
+    _select_rotation_angles method. Then, we can use the fact that E SO(4) Edag
+    gives us something in SU(2) x SU(2) to give A, B, C, D.
+    """
+
+    # A lot of the work here happens in the magic basis. Essentially, we
+    # don't look explicitly at some U = G V H, but rather at
+    #     E^\dagger U E = G E^\dagger V E H
+    # so that we can recover
+    #     U = (E G E^\dagger) V (E H E^\dagger) = (A \otimes B) V (C \otimes D).
+    # There is some math in the paper explaining how when we define U in this way,
+    # we can simultaneously diagonalize functions of U and V to ensure they are
+    # in the same coset and recover the decomposition.
+
+    u = math.dot(math.cast_like(Edag, V), math.dot(U, math.cast_like(E, V)))
+    v = math.dot(math.cast_like(Edag, V), math.dot(V, math.cast_like(E, V)))
+
+    uuT = math.dot(u, math.T(u))
+    vvT = math.dot(v, math.T(v))
+
+    # First, we find a matrix p (hopefully) in SO(4) s.t. p^T u u^T p is diagonal.
+    # Since uuT is complex and symmetric, both its real / imag parts share a set
+    # of real-valued eigenvectors.
+    if qml.math.get_interface(u) == "tensorflow":
```

Comment:
This was the simplest way to get things up and running in all cases. If the regular version is used, tensorflow does not produce purely real eigenvectors.

Reply:
Is there a reason for not just using
```python
ev_p, p = math.linalg.eig(qml.math.real(uuT))
```
for all interfaces?

Reply:
Doing that in tensorflow produced complex eigenvectors. They are valid eigenvectors, but we need things to be in SO(4) here. This didn't seem to be a problem in the other interfaces; however when I switched to using the real part here instead, I ran into separate issues so separating the cases was the simplest option for the time being to get the 3-CNOT case to work :sad: